### DISCUSSION TITLE
Join Collection, Basic Operations  and JSON Storage

### DISCUSSION THREAD

**Discussion by iamdineshkumar** [appwrite/appwrite#2901] [Q&A]
1. Is that possible to join collections and get data?
2. Whether storage of JSON object is allowed?
3. Is that possible to implement arithmetic operations like 
- sum of field 
- avg of field
- Distinct value of field
- count
- min & Max

**Comment by gewenyu99** [comment]
For context, I've also seen various requests for the following in different contexts and places:
- Querying JSON objects (which needs json type objects to exist first)
- Atomic transactions, arithmetic, array updates
- Array operations
- Array querying

Putting these here so we can keep track of them.

**Reply by maeddin** [reply]
What does querying for JSON objects mean?  Are these equal to embedded objects because you can query for the attributes of the JSON objects?

**Reply by gewenyu99** [reply]
Yep. I've seen requests to basically support embedded objects.

While I have no idea if this will/should happen. I am adding these suggestions here to further the discussion.

**Reply by maeddin** [reply]
If that were the way to go, it would hardly make sense to take MariaDB as the backend. Unless you want to allow embedded objects and joins.

**Reply by iamdineshkumar** [reply]
@MaddinMade @gewenyu99  I have been working in MongoDB for years... Really it give lot of flexibility interms of handling json objects. Right now i created cloud functions to handle same and using mongoDB to handle those stuffs while i use remaning services of appwrite. But for a long run it would be great if its default in appwrite. For your additional information Open run time of PHP not support mongodb due to various reason like its server environment is not supporting. So now i am using node js for cloud function environment to handle mongodb.

**Reply by maeddin** [reply]
@iamdineshkumar Yes, this is also the reason why I use MongoDB for new projects. The object orientation allow quick adjustments to the structure if you want to make small changes. With relational or other no-SQL databases, everything has to be better thought out for the most part if you don't want to do major migrations later. This is also the reason why I think object orientation would make more sense for AppWrite than SQL. And when you deal with documents, MongoDB is simply much more performant than any MariaDB wrapper. With joins, on the other hand, MariaDB is of course faster, even if MongoDB now also has a join alternative, if they are ever necessary.

But I did not know that MongoDB makes problems with PHP. I don't work with PHP on my own projects. 😅

**Reply by maeddin** [reply]
And to be fair: MongoDB is very fast for an ACID database.

**Reply by iamdineshkumar** [reply]
> 
Actually open runtime of php is using [Swoole](https://github.com/swoole/swoole-src) as server which has issues with mongodb it seems. But some came up with custom build for handling mongodb but which is not officially available to appwrite so i switched cloud function running environment to node js. Now i am using mongodb for database and rest through appwrite kind of hybrid :)
